File: 567-permutation-in-string/permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if(s1.length()>s2.length()){
            return false;
        }
        unordered_map<char,int> mp1;
        for(char ch:s1){
            mp1[ch]++;
        }
        int left=0;
        unordered_map<char,int> mp2;
        for(int right=0;right<s2.length();right++){
            mp2[s2[right]]++;
            while(right-left+1>s1.length()){
                mp2[s2[left]]--;
                if(mp2[s2[left]]==0){
                    mp2.erase(s2[left]);
                }
                left++;
            }
            if(mp1==mp2){
                return true;
            }
        }
        return false;
    }
};

File: 567-permutation-in-string/permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if(s1.length()>s2.length()){
            return false;
        }
        HashMap<Character,Integer> mp1=new HashMap<>();
        for(char ch:s1.toCharArray()){
            mp1.put(ch,mp1.getOrDefault(ch,0)+1);
        }
        int left=0;
        HashMap<Character,Integer> mp2=new HashMap<>();
        for(int right=0;right<s2.length();right++){
            mp2.put(s2.charAt(right),mp2.getOrDefault(s2.charAt(right),0)+1);
            while(right-left+1>s1.length()){
                mp2.put(s2.charAt(left),mp2.get(s2.charAt(left))-1);
                if(mp2.get(s2.charAt(left))==0){
                    mp2.remove(s2.charAt(left));
                }
                left++;
            }
            if(mp1.equals(mp2)){
                return true;
            }
        }
        return false;
    }
}